File: python/construct_binary_tree_from_preorder_and_inorder_traversal_105/Solution.py
from typing import List, Optional

from tree.binary_tree.TreeNode import TreeNode


class Solution:
    def buildTree(self, preorder: List[int], inorder: List[int]) -> Optional[TreeNode]:
        if not inorder:
            return None

        num: int = preorder.pop(0)
        index: int = inorder.index(num)
        root = TreeNode(num)

        root.left = self.buildTree(preorder, inorder[:index])
        root.right = self.buildTree(preorder, inorder[index + 1 :])
        return root

File: java/construct_binary_tree_from_preorder_and_inorder_traversal_105/Solution.java
package construct_binary_tree_from_preorder_and_inorder_traversal_105;

import libraries.tree.binary_tree.TreeNode;

import java.util.ArrayDeque;
import java.util.Arrays;
import java.util.Queue;

class Solution {
  public TreeNode buildTree(int[] preorder, int[] inorder) {
    Queue<Integer> preorderQueue = new ArrayDeque<>();
    for (int num : preorder) {
      preorderQueue.add(num);
    }

    return build(preorderQueue, inorder);
  }

  private TreeNode build(Queue<Integer> preorder, int[] inorder) {
    if (inorder.length == 0) {
      return null;
    }

    int num = preorder.poll();
    int index = indexOf(inorder, num);
    TreeNode root = new TreeNode(num);

    root.left = build(preorder, Arrays.copyOfRange(inorder, 0, index));
    root.right = build(preorder, Arrays.copyOfRange(inorder, index + 1, inorder.length));

    return root;
  }

  private static int indexOf(int[] array, int target) {
    for (int i = 0; i < array.length; i++) {
      if (array[i] == target) {
        return i;
      }
    }
    return -1;
  }
}
